src/middlewared/middlewared/plugins/directoryservices_/util_cache.py

```diff
@@ -0,0 +1,368 @@
+        try:
+            if self.users_handle:
+                self.users_handle.close()
+        except Exception as exc:
+            stored_exception = exc
+
+        try:
+            if self.groups_handle:
+                self.groups_handle.close()
+        except Exception as exc:
+            stored_exception = exc
+
+        if stored_exception:
+            raise stored_exception
+
+    def _commit(self):
+        os.rename(self.users_handle.full_path, DSCacheFile.USER.path)
+        os.rename(self.groups_handle.full_path, DSCacheFile.GROUP.path)
+
+    def _add_sid_info_to_entries(
+        self,
+        idmap_ctx: idmap_winbind.WBClient | idmap_sss.SSSClient,
+        nss_entries: list,
+        dom_by_sid: dict
+    ) -> list:
+        """
+        Add SID information to entries that NSS has returned. Dictionary
+        entries in list `nss_entries` are modified in-place.
+
+        `idmap_ctx` - is the winbind or sssd client handle to use to resolve
+        posix accounts to SIDs
+
+        `nss_entries` - list of posix accounts to look up
+
+        `dom_by_sid` - mapping for trusted domains to provide idmap backend
+        information for trusted domains. This is used to ensure that synthetic
+        database IDs are unique and guaranteed to not change.
+
+        Returns:
+            Same list passed in as nss_entries
+        """
+        idmaps = idmap_ctx.users_and_groups_to_idmap_entries(nss_entries)
+        to_remove = []
+
+        for idx, entry in enumerate(nss_entries):
+            unixkey = f'{IDType[entry["id_type"]].wbc_str()}:{entry["id"]}'
+            if unixkey not in idmaps['mapped']:
+                # not all users / groups in SSSD have SIDs
+                # and so we'll leave them with a null SID
+                # rather than removing from nss_entries
+                continue
+
+            idmap_entry = idmaps['mapped'][unixkey]
+            if idmap_entry['sid'].startswith((SID_LOCAL_GROUP_PREFIX, SID_LOCAL_USER_PREFIX)):
+                # There is a collision between local user / group and our AD one.
+                # pop from cache
+                to_remove.append(idx)
+                continue
+
+            if idmap_entry['sid'].startswith(SID_BUILTIN_PREFIX):
+                # We don't want users to select auto-generated builtin groups
+                to_remove.append(idx)
+                continue
+
+            entry['sid'] = idmap_entry['sid']
+            entry['id_type'] = idmap_entry['id_type']
+            if dom_by_sid:
+                entry['domain_info'] = dom_by_sid[idmap_entry['sid'].rsplit('-', 1)[0]]
+
+        to_remove.reverse()
+        for idx in to_remove:
+            nss_entries.pop(idx)
+
+        return nss_entries
+
+    def _get_entries_for_cache(
+        self,
+        idmap_ctx: idmap_winbind.WBClient | idmap_sss.SSSClient | None,
+        nss_module: NssModule,
+        entry_type: IDType,
+        dom_by_sid: dict
+    ) -> list:
```

Comment:
Adding type hints and return types gives more opportunity for Flake8 to find things to complain about.

> Return type of generator function must be compatible with "Generator[Any, Any, Any]"
>   "Generator[Any, Any, Any]" is incompatible with "list"